src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,483 @@
+        3. make a log record if the DID + Rucio account tuple does not have any
+        existent rules
+           - and set ruleIds and currentRSEs to an empty list
+        4. once all the relevant rules have been removed, persist an up-to-date
+        version of the pileup data structure in MongoDB
+        """
+        spec = {'active': False}
+        docs = self.mgr.getPileup(spec)
+        taskSpec = self.getTaskSpec()
+        self.logger.info("Running the inactive task on %d pileup objects", len(docs))
+        asyncio.run(performTasks('inactive', docs, taskSpec))
+
+    def activeTask(self, nodeUsage=None, marginSpace=1024**4):
+        """
+        Active pileup task:
+        :param nodeUsage: node usage dictionary, see RSEQuotas
+        :param marginSpace: minimum margin space size in bytes to have at RSE to place a dataset
+
+        This task is supposed to look at pileup documents active in the system.
+        Its main goal is to ensure that the pileup DID has all the requested
+        rules (and nothing beyond them), according to the pileup object
+        configuration. Pileup documents that are updated as a result of this
+        logic should have their data persisted back in MongoDB.
+
+        1. Read pileup document from MongoDB with filter active=true
+        2. if expectedRSEs is different than currentRSEs, then further data placement
+        is required (it's possible that data removal is required!)
+        3. make a local copy of the currentRSEs value to track rules incomplete but ongoing
+        4. for each rule matching the DID + Rucio account, perform:
+           - if rule RSE is not in expectedRSEs, then this rule needs to be deleted.
+             - upon successful rule deletion, also remove the RSE name
+             from currentRSEs and ruleIds (if any)
+             - make a log record
+           - else, save this rule RSE in the local copy of currentRSEs.
+           This rule is likely still being processed by Rucio.
+        5. now that we evaluated expected versus current,
+        for each expectedRSEs not in our local copy of currentRSEs:
+           - first, check whether the RSE has enough available space available for that
+           (we can assume that any storage with less than 1 TB available cannot be
+           considered for pileup data placement)
+             - in case of no space available, make a log record
+           - in case there is enough space, make a Rucio rule for
+           that DID + Rucio account + RSE
+             - now append the rule id to the ruleIds list
+        6. once all the relevant rules have been created,
+        or if there was any changes to the pileup object,
+        persist an up-to-date version of the pileup data structure in MongoDB
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        taskSpec = self.getTaskSpec()
+        taskSpec['marginSpace'] = marginSpace
+        taskSpec['nodeUsage'] = nodeUsage
```

Comment:
Given that each co-routine is an independent work unit, it could be that they are dealing with the quota based on an outdated value. In other words, it could be that within the same cycle, we are placing many datasets under the same RSE and exhausting (or going beyond) their actual quota.

This needs to be revisited in the upcoming PR.

Reply:
If this is the case, then we need to make rucio RSE usage call for every RSE each time. So far I made one rucio call to get all RSE usage info. Please advise which way it should be implemented.